### ISSUE TITLE
iCal (ics) support

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#1580]
Originally reported on Google Code with ID 1313

```
Would be great if we could import ics files or if there was some documentation on how
to do so.
```

Reported by `andrewmcmurry` on 2012-02-22 23:57:06

Imported with **19** stars.

**Comment by 2733** [comment]
Hi, i wrote this https://github.com/2733/icalendar2fullcalendar.
Dunno if it could be useful to integrate it in FullCalendar.

**Comment by kingcheez** [comment]
Thanks. Does it work for recurring events?
 
On Thu, Mar 10, 2016, at 10:36 PM, 2733 wrote:
> Hi, i wrote this https://github.com/2733/icalendar2fullcalendar.
>
Dunno if it could be useful to integrate it in FullCalendar.
> —
> Reply to this email directly or view it on GitHub[1].
>
 

Links:

  1. https://github.com/fullcalendar/fullcalendar/issues/1580#issuecomment-195164076

**Comment by 2733** [comment]
No, because fullcalendar doesnt support recurring events, does it?
I can workaround that by generating one-time events from a recurring event.
Any advice?

**Comment by kingcheez** [comment]
Not exactly, but ical.js exposes an API for evaluating recurrences over a period:
 
https://mozilla-comm.github.io/ical.js/api/ICAL.RecurExpansion.html
 
This could be used in combination with the start/end parameters from the
'events' callback in fullcalendar and made to work.

**Comment by 2733** [comment]
Now recurring events work too.

**Comment by kingcheez** [comment]
This seems overly complex (two interface functions when I think only one is needed) but nice to see that it works. I might have an opportunity to test it later. Thanks for doing the work!

On March 16, 2016 12:01:18 PM EDT, 2733 <notifications@github.com> wrote:
>Now recurring events work too.
>
>---
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/fullcalendar/fullcalendar/issues/1580#issuecomment-197397629

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by 2733** [comment]
Please report this and other issues on my repo, not to go off-topic here.

**Comment by DoobleD** [comment]
Hi, any news on this new feature ? Not implemented yet ?

**Comment by pokoli** [comment]
@ghost do you plan to create a PR on this repo to integrate it as base feature?

**Comment by larrybolt** [comment]
If you need this, apparently it's not so difficult to do yourself using [ical.js](https://github.com/mozilla-comm/ical.js), demo [here](https://github.com/larrybolt/online-ics-feed-viewer/blob/master/index.html#L23). It's not super clearly documented and nice to work with, but good enough if you really need this.

**Comment by ace280** [comment]
tl;dr: Why not use the naming scheme defined in RFC 5545 to denote calendar and event elements in FullCalendar?

I am currently trying to implement FullCalendar in [Redmine](https://www.redmine.org/). I am to provide the same functionality their calendar already has and in addition the calendar data are to be saved on a calendar server such as [Radicale](https://radicale.org/). The main benefit shall be to enable synchronizing with other calendar clients such as [Evolution](http://wiki.gnome.org/Apps/Evolution/), Thunderbird, or Microsoft Office Outlook.

So far, integrating FullCalendar works well, but using the ics-data received from Evolution is rather time consuming. In general I agree that FullCallendar accepting data in the iCal (RFC 5545) or jCal (RFC 7265) format would be best. 

**What would help a lot already would be if FullCalendar used the naming conventions specified in RFC 5545**. This would make the mapping of iCal Data in the back-end much easier. Users would not have to learn the way, FullCalendar names things compared to how iCal names things. Some  Examples: 
- why is it uid in iCal and id in Fullcalendar? 
- Why is it start for non recurring events and dtstart for the recurrence rule while in iCal this is the same value, both denoted with dtstart?

In the same way dtstart is not included in the iCal rrule, but it has to be included in the Fullcalendar rrule. I believe that providing the Event Parsing could become so much easier if the naming scheme was a one to one mapping. 

If this was achieved, native data formats of whatever source could be parsed to JSON and used in FullCalendar. In my case I am using the ruby implementation of [Icalendar](https://github.com/icalendar/icalendar) which saves event data in hashes internally. These, in turn, could easily be rendered as JSON. But with the current FullCalendar implementation I will first have to map the iCal identifiers for the events to the FullCalendar identifiers and handle the rrule (which has proven rather cumbersome). 

With this explanation I hope to outline a way that would make things much easier without having to add an iCal Parser or an own CalDav server such as suggested in https://github.com/fullcalendar/fullcalendar/issues/690. Please tell me if something is unclear as English is not my native language.

Thank you for the great work. I will return to writing a Icalendar to FullCalendar parser in Ruby now ;)

**Comment by ace280** [comment]
> Now recurring events work too.

I had a look into it, and it seems pretty promising. Maybe the [open pull request](https://github.com/leonaard/icalendar2fullcalendar/pull/5) might make it even easier to integrate this into FullCalendar? 

> Please report this and other issues on my repo, not to go off-topic here.

I think that this is the right place to discuss your work, because it could be the foundation for integrating ICS-support without having to break the current naming scheme.

**Comment by ace280** [comment]
This repository seems to have done most of the job as well: https://github.com/larrybolt/online-ics-feed-viewer

**Comment by sigalou** [comment]
Hello,

I went through all your documentation of all your sites and I am wondering about the lack of consideration of the ICS protocol. Which seems to me to be the universal format of calendars.

You really favor Google Calendar (https://fullcalendar.io/docs/google-calendar) and I can't find anything on your site that natively allows the integration of an ICS link.
Yet discussions of several years back evoked ICS (#1580) then nothing since, this idea has been totally abandoned?
My research led me to:

    https://github.com/leonaard/icalendar2fullcalendar
    https://github.com/larrybolt/online-ics-feed-viewer (even online version does not work)
    Extensions that use https://mozilla-comm.github.io/ical.js/ to fetch ical and send it to FullCalendar.

Do you think it's useful to include a link, natively, in FullCalendar which could retrieve a URL in iCal?
This idea is completely abandoned and each one needs to develop their own interface?
Can we work on it together?

Thank you, and good continuation on this project, still relevant.

**Comment by Clay-Ferguson** [comment]
+1. Not to be critical of the developers of this great component, but it seems to me almost unfathomable that the best calendar app out there right now (FullCalendar) doesn't even support what has been the universal file (ICS) format for all calendar apps for over a decade now, despite the ICS format being relatively simple.

**Comment by pabrams** [comment]
Agree that it's quite simple to implement this. I did it myself in about a day or so.  But I did it for GoC/Immigration, so I don't think I'm allowed to contribute it here.

**Comment by arshaw** [comment]
This has been implemented in v5.5.0 with the `@fullcalendar/icalendar` package, which internally uses the [ical.js project](https://github.com/mozilla-comm/ical.js/).

Docs: https://fullcalendar.io/docs/icalendar

This implementation is intentionally simple. If you need anything beyond this, it's probably falls into the realm of CalDAV, which [there's another ticket for](https://github.com/fullcalendar/fullcalendar/issues/690).

Please post post bug reports and feature requests in new tickets separate from this one. Thanks.
